README.md
resque-exceptional
==================

resque-exceptional provides a [Resque][re] failure backend that sends exceptions
raised by jobs to [getexceptional.com][ge]

[![Build Status](https://secure.travis-ci.org/lantins/resque-exceptional.png?branch=master)](http://travis-ci.org/lantins/resque-exceptional)

Install & Quick Start
---------------------

Before you jump into code, you'll need a getexceptional.com account.

To install:

    $ gem install resque-exceptional

### Example: Single Failure Backend

Using only the exceptional failure backend:

    require 'resque'
    require 'resque-exceptional'

    Resque::Failure::Exceptional.configure do |config|
      config.api_key = '505f2518c41866bb0be7ba434bb2b079'
      config.use_ssl = false
    end

    Resque::Failure.backend = Resque::Failure::Exceptional

### Example: Multiple Failure Backends

Using both the redis and exceptional failure backends:

    require 'resque'
    require 'resque-exceptional'

    require 'resque/failure/multiple'
    require 'resque/failure/redis'

    Resque::Failure::Exceptional.configure do |config|
      config.api_key = '505f2518c41866bb0be7ba434bb2b079'
    end

    Resque::Failure::Multiple.classes = [Resque::Failure::Redis, Resque::Failure::Exceptional]
    Resque::Failure.backend = Resque::Failure::Multiple

Configuration Options
---------------------

**Required**

  * `api_key` - your getexceptional.com api key.

**General Options** *(optional)*

  * `deliver` - set `false` to disable delivery of errors to exceptional API, handy for testing (default: `true`)

**HTTP Proxy Options** *(optional)*

  * `proxy_host` - proxy server ip / hostname.
  * `proxy_port` - proxy server port.
  * `proxy_user` - proxy server username.
  * `proxy_pass` - proxy server password.

**HTTP Client Options** *(optional)*

  * `use_ssl` - set `true` if your plan supports ssl. (default: `false`)
  * `http_open_timeout` - timeout in seconds to establish the connection. (default: `2`)
  * `http_read_timeout` - timeout in seconds to wait for a reply. (default: `5`)

Note on Patches/Pull Requests
-----------------------------

  * Fork the project.
  * Make your feature addition or bug fix.
  * Add tests for it. This is important so I don't break it in a future
    version unintentionally.
  * Commit, do not mess with the version. (if you want to have your own
    version, that is fine but bump version in a commit by itself I can ignore
    when I pull)
  * Send me a pull request. Bonus points for topic branches.

Author
------

Luke Antins :: [http://lividpenguin.com][lp] :: @lantins

[re]: http://github.com/defunkt/resque
[lp]: http://lividpenguin.com
[ge]: http://getexceptional.com

HISTORY.md
## 0.2.2 (2014-02-21)

* Add `deliver?` helper method.
* Improve tests and ensure they are not order order.

## 0.2.1 (2014-02-21)

* Add `deliver` config option; `false` will disable delivery, good for tests.

## 0.2.0 (2012-09-07)

* Changed API URL to: api.exceptional.io
* Changed test framework to MiniTest.
* Added Gemfile.
* Replaced usage of `#to_json` with `MultiJson.dump()`

## 0.1.0 (2010-10-16)

* Added more unit tests.
* Added yardoc code comments.
* Had a couple days production testing ;-) Initial version to be announced.

## 0.0.1 (2010-10-13)

* Initial release.

test/resque_test.rb
require 'test_helper'

# make sure the worlds not fallen from beneith us.
class ResqueTest < Minitest::Test
  def test_resque_version
    major, minor, patch = Resque::Version.split('.')
    assert_equal 1, major.to_i, 'major version does not match'
    assert_operator minor.to_i, :>=, 8, 'minor version is too low'
  end

end


test/test_helper.rb
# fix load path.
dir = File.dirname(File.expand_path(__FILE__))
$LOAD_PATH.unshift dir + '/../lib'
$TESTING = true

# Run code coverage in MRI 1.9 only.
if RUBY_VERSION >= '1.9' && RUBY_ENGINE == 'ruby'
  require 'simplecov'
  SimpleCov.start do
    add_filter '/test/'
  end
end

# require gems for testing.
require 'minitest/autorun'
require 'minitest/pride'
require 'rr'
require 'webmock'
require 'webmock/minitest'

# require our failure backend to test.
require 'resque-exceptional'

class Minitest::Test
  include WebMock::API

  # periodicly set the api key.
  def with_api_key(key, &block)
    Resque::Failure::Exceptional.api_key = key
    yield
    Resque::Failure::Exceptional.api_key = nil
  end
end

# fake worker.
class FakeWorker
  attr_reader :log_history

  def initialize
    @log_history = []
  end

  def log(msg)
    @log_history << msg
    p msg if ENV['VERBOSE']
  end

  def to_s
    'mr. fake resque worker.'
  end
end

# test exceptions.
module TestApp
  class Error < StandardError
  end

  def self.method_bar
    raise Error, 'example exception message. bar.'
  end

  def self.method_foo
    method_bar
  end

  def self.grab_exception
    begin
      method_foo
    rescue => e
      return e
    end
  end

end


test/exceptional_test.rb
require 'test_helper'

# Tests the failure backend works with resque, does not contact the api.
class ExceptionalTest < Minitest::Test
  def setup
    @exception = TestApp.grab_exception
    @worker = FakeWorker.new
    @queue = 'test_queue'
    @payload = { 'class' => 'TestJob', 'args' => ['foo', 'bar'] }
    @failure = Resque::Failure::Exceptional.new(@exception, @worker, @queue, @payload)
    WebMock.disable_net_connect!
    WebMock.reset!
  end

  # test we can build a hash to send to the api.
  def test_can_build_api_request_data_hash
    data = @failure.api_request
    assert_kind_of Hash, data, 'should build a hash'
  end

  # include the minimum entries required by the api.
  def test_api_request_includes_minimum_api_entries
    data = @failure.api_request

    assert_kind_of Hash,   data['application_environment']
    assert_kind_of Hash,   data['application_environment']['env']
    assert_kind_of String, data['application_environment']['application_root_directory']

    assert_kind_of Hash, data['exception']
    # test occurred_at that starts like: 2010-10-13T01:56:49
    assert_match /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/, data['exception']['occurred_at']
    assert_kind_of String, data['exception']['message']
    assert_kind_of Array, data['exception']['backtrace']
    assert_equal 'TestApp::Error', data['exception']['exception_class']
  end

  # include the resque information, we sneak it in as request data so we can
  # view it via the web interface.
  def test_api_request_includes_resque_info
    data = @failure.api_request

    assert_kind_of Hash,   data['request']
    assert_kind_of Hash,   data['request']['parameters']
    assert_kind_of String, data['request']['parameters']['queue']
    assert_kind_of String, data['request']['parameters']['job_class']
    assert_kind_of Array,  data['request']['parameters']['job_args']
    assert_kind_of String, data['request']['parameters']['worker']
  end

  # make it more obvious in the web interface this was a resque failure.
  def test_api_request_includes_resque_block
    data = @failure.api_request

    assert_kind_of Hash, data['rescue_block']
    assert_equal 'Resque Failure', data['rescue_block']['name']
  end

  # let them know who we are.
  def test_api_request_includes_client_info
    data = @failure.api_request

    assert_equal 'resque-exceptional', data['client']['name'], 'should use the gem name'
    assert_match /^\d+\.\d+\.\d+$/, data['client']['version'], 'should have a version number'
    assert_kind_of Fixnum, data['client']['protocol_version']
  end

  # we need the ability to configure the failure backend before its created.
  # config settings should be class variables.
  def test_configure
    Resque::Failure::Exceptional.configure do |config|
      config.api_key = 'my api key.'
      config.deliver = false
      # everything below are http client options.
      config.proxy_host = 'host.name.com'
      config.proxy_port = 8080
      config.proxy_user = 'foo'
      config.proxy_pass = 'bar'
      config.use_ssl = true
      config.http_open_timeout = 5
      config.http_read_timeout = 10
    end

    # reset everything to nil...
    Resque::Failure::Exceptional.configure do |config|
      options = %w{api_key deliver proxy_host proxy_port proxy_user proxy_pass
                   use_ssl http_open_timeout http_read_timeout}
      options.each { |opt| config.send("#{opt}=", nil) }
    end
  end

  # failure backends need to define a save method.
  def test_save_defined
    assert_equal true, @failure.respond_to?(:save)
  end

  # test save success.
  def test_save_success
    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      stub_request(:post, /.*api.exceptional.io.*/)
      @failure.save
      assert_requested(:post, /.*api.exceptional.io.*/)
      assert_match /^(resque-exception).*(success).*$/, @worker.log_history.first
    end
  end

  # test save fail.
  def test_save_fail
    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      stub_request(:post, /.*api.exceptional.io.*/).to_return(:status => [500, 'Internal Server Error'])
      @failure.save
      assert_requested(:post, /.*api.exceptional.io.*/)
      assert_match /^(resque-exception).*(fail).*(500).*$/, @worker.log_history.first
    end
  end

  # test save does not submit to Exceptional if deliveries are disabled
  def test_do_not_submit_if_deliveries_are_disabled
    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      # turn off delivery.
      Resque::Failure::Exceptional.configure { |c| c.deliver = false }
      @failure.save
      assert_equal 0, @worker.log_history.length, 'should not send any requests'

      # Put default setting back in place.
      Resque::Failure::Exceptional.configure { |c| c.deliver = nil }
    end
  end

  # we need a Net::HTTP client setup to send the data.
  def test_http_client
    assert_kind_of Net::HTTP, @failure.http_client
  end

  # we should fail if the api_key is not set.
  def test_fail_if_api_key_nil
    # should already be nil, but lets just be sure...
    with_api_key nil do
      @failure.save
      assert_match /^(resque-exception).*(error).*(api_key).*$/, @worker.log_history.first
    end
  end

  # test we prefix our log messages.
  def test_log_adds_prefix
    @failure.log('test message')
    @failure.log('123 another msg bud!')
    assert_match /^resque-exception - .*/, @worker.log_history.first
    assert_match /^resque-exception - .*/, @worker.log_history.last
  end

  # test our `#use_ssl?`, `deliver?` and `#http_port` helper methods.
  def test_helper_methods
    # enable ssl.
    Resque::Failure::Exceptional.configure { |c| c.use_ssl = true }
    assert_equal true, @failure.use_ssl?, 'use_ssl? should now be true'
    assert_equal 443, @failure.http_port, 'http_port should now be 443.'

    # disable ssl.
    Resque::Failure::Exceptional.configure { |c| c.use_ssl = nil }
    assert_equal false, @failure.use_ssl?, 'use_ssl? should default to false.'
    assert_equal 80, @failure.http_port, 'http_port should default to 80.'

    # disable delivery.
    Resque::Failure::Exceptional.configure { |c| c.deliver = false }
    assert_equal false, @failure.deliver?, 'deliver should now be false'

    # enable delivery.
    Resque::Failure::Exceptional.configure { |c| c.deliver = nil }
    assert_equal true, @failure.deliver?, 'deliver should now be true'
  end

  # returns nil if the backtrace is empty.
  def test_uniqueness_hash_returns_nil_when_empty_backtrace
    mock(@failure.exception).backtrace.times(any_times) { Array.new }
    assert_equal nil, @failure.uniqueness_hash
  end

  # returns nil if the backtrace is empty.
  def test_uniqueness_hash_returns_nil_when_nil_backtrace
    mock(@failure.exception).backtrace.times(any_times) { nil }
    assert_equal nil, @failure.uniqueness_hash
  end

  # uniqueness_hash builds a md5sum.
  def test_uniqueness_hash_returns_a_md5_of_the_backtrace
    # fake backtrace.
    fake_backtrace = ['fake', 'backtrace', 'that_wont_change']
    mock(@failure.exception).backtrace.times(any_times) { fake_backtrace }

    assert_equal '27810b263f0e11eef2f1d29be75d2f39', @failure.uniqueness_hash
  end

  # return the HTTP path and query string with uniqueness hash.
  def test_http_path_query
    # fake backtrace.
    fake_backtrace = ['fake', 'backtrace', 'that_wont_change']
    mock(@failure.exception).backtrace.times(any_times) { fake_backtrace }

    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      path, query = *@failure.http_path_query.split('?', 2)
      assert_match /^api_key=27810b263f0e11eef2f1d29be75d2f39/, query, 'query should have api_key.'
      assert_match /protocol_version=\d{1}/, query, 'query should have protocol_version.'
      assert_match /hash=27810b263f0e11eef2f1d29be75d2f39$/, query, 'query should have a uniqueness hash.'
    end
  end

  # build a path & query without a uniqueness hash.
  def test_http_path_query_without_uniqueness_hash
    # fake empty backtrace.
    mock(@failure.exception).backtrace.times(any_times) { Array.new }

    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      path, query = *@failure.http_path_query.split('?', 2)
      assert_match /^api_key=27810b263f0e11eef2f1d29be75d2f39/, query, 'query should have api_key.'
      assert_match /protocol_version=\d{1}$/, query, 'query should have protocol_version.'
    end
  end

  # raise exception if api key is not set.
  def test_http_path_query_without_api_key_raises_exception
    with_api_key nil do
      assert_raises Resque::Failure::Exceptional::APIKeyError, 'should raise APIKeyError if api key is not set' do
        @failure.http_path_query
      end
    end
  end

  # should return http response if successful.
  def test_http_post_request
    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      stub_request(:post, /.*api.exceptional.io.*/)

      response = @failure.http_post_request
      assert_requested(:post, /.*api.exceptional.io.*/)
      assert_equal '200', response.code, 'should be a successful http request'
    end
  end

  # should handle exceptions raised during the HTTP Post.
  def test_http_post_request_handles_exceptions_and_returns_nil
    response = @failure.http_post_request
    assert_equal nil, @failure.http_post_request, 'should be nil, APIKeyError should have been caught.'

    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      WebMock.reset!
      stub_request(:post, /.*api.exceptional.io.*/).to_raise(StandardError)
      assert_equal nil, @failure.http_post_request, 'should be nil, StandardError should have been caught.'
      assert_requested(:post, /.*api.exceptional.io.*/)
    end
  end

  # make sure we catch timeout errors.
  def test_http_post_request_timeout
    with_api_key '27810b263f0e11eef2f1d29be75d2f39' do
      stub_request(:post, /.*api.exceptional.io.*/).to_timeout
      assert_equal nil, @failure.http_post_request, 'should be nil, TimeoutError should have been caught.'
      assert_requested(:post, /.*api.exceptional.io.*/)
    end
  end

  # perform a test with the real api.
  def test_live_fire_with_real_api!
    unless ENV['EXCEPTIONAL_API_KEY']
      skip 'Test with the REAL API. Example: `EXCEPTIONAL_API_KEY=27810b263f0e11eef2f1d29be75d2f39 bundle exec rake test`'
    end

    with_api_key ENV['EXCEPTIONAL_API_KEY'] do
      WebMock.allow_net_connect!
      @failure.save
      assert_match /^(resque-exception).*(success).*$/, @worker.log_history.first
    end
  end

end


lib/resque/failure/exceptional.rb
module Resque
  module Failure
    # A Resque failure backend that sends exception data to getexceptional.com
    class Exceptional < Base
      Version = '0.2.2' # Failure backend version number.

      # Raised if the api_key is not set.
      class APIKeyError < StandardError
      end

      class << self
        attr_accessor :api_key # your getexceptional api key.
        attr_accessor :use_ssl # enable/disable SSL.
        attr_accessor :deliver # whether or not to submit exceptions to Exceptional
        # HTTP proxy option
        attr_accessor :proxy_host, :proxy_port, :proxy_user, :proxy_pass
        # HTTP client option
        attr_accessor :http_open_timeout, :http_read_timeout
      end

      # Configures the failure backend. At a minimum you will need to set
      # an api_key.
      #
      # @example Setting your API Key and enabling SSL:
      #   Resque::Failure::Exceptional.configure do |config|
      #     config.api_key = '505f2518c41866bb0be7ba434bb2b079'
      #     config.use_ssl = true
      #   end
      def self.configure
        yield self
      end

      # Sends the exception data to the exceptional api.
      #
      # When a job fails, a new instance is created and #save is called.
      def save
        return unless deliver?
        return unless response = http_post_request

        if response.code == '200'
          log "success - api accepted the exception data."
        else
          body = response.body if response.respond_to? :body
          log "fail - expected: 200 OK, received: #{response.code} #{response.message}"
        end
      end

      # Sends a HTTP Post to the exceptional api.
      #
      # @return [Net::HTTPResponse] http response data.
      # @return [nil] if something went wrong.
      def http_post_request
        begin
          return http_client.post(http_path_query, compressed_request, http_headers)
        rescue APIKeyError
          log 'error - you must set your api_key.'
        rescue TimeoutError
          log 'fail - timeout while contacting the api server.'
        rescue Exception => e
          log "fail - exception raised during http post. (#{e.class.name}: #{e.message})"
        end
        nil
      end

      # HTTP headers to send.
      #
      # @return [Hash] http headers.
      def http_headers
        {
          'Content-Type' => 'application/json',
          'Accept'       => 'application/json',
          'User-Agent'   => "resque-exceptional/#{Version}"
        }
      end

      # Returns the compressed request data.
      def compressed_request
        Zlib::Deflate.deflate(MultiJson.dump(api_request), Zlib::BEST_SPEED)
      end

      # Path & query options used by the HTTP Post.
      #
      # @raise [APIKeyError] if the api_key is not set.
      # @return [String] http path & query options.
      def http_path_query
        raise APIKeyError, 'api key must be set.' unless self.class.api_key
        hash_param = uniqueness_hash.nil? ? nil : "&hash=#{uniqueness_hash}"
        "/api/errors?api_key=#{self.class.api_key}&protocol_version=5#{hash_param}"
      end

      # Calculates a uniqueness md5sum of the exception backtrace if available.
      #
      # nb. this isn't documented in the public api... not sure if we should
      # use it or not...
      #
      # @return [String] md5sum of the backtrace.
      # @return [nil] if we don't have a backtrace available.
      def uniqueness_hash
        return nil if (exception.backtrace.nil? || exception.backtrace.empty?)
        Digest::MD5.hexdigest(exception.backtrace.join)
      end

      # Configures a HTTP client.
      #
      # @return [Net::HTTP] http client.
      def http_client
        # pass any proxy settings.
        proxy = Net::HTTP::Proxy(self.class.proxy_host, self.class.proxy_port,
                                 self.class.proxy_user, self.class.proxy_pass)
        http = proxy.new('api.exceptional.io', http_port)

        # set http client options.
        http.read_timeout = self.class.http_read_timeout || 5
        http.open_timeout = self.class.http_open_timeout || 2
        http.use_ssl = use_ssl?

        http
      end

      # Helper method to return the correct HTTP port number, depending on
      # if were using SSL or not.
      #
      # @return [Fixnum] HTTP port number.
      def http_port
        use_ssl? ? 443 : 80
      end

      # Helper method to check if were using SSL or not.
      #
      # @return [Boolean] true if ssl is enabled.
      def use_ssl?
        self.class.use_ssl || false
      end

      # Helper method to check if errors should be submitted to exceptional API.
      #
      # @return [Boolean] true if deliver is enabled.
      def deliver?
        self.class.deliver.nil? || self.class.deliver
      end

      # Adds a prefix to log messages.
      #
      # @param [String] msg your log message.
      def log(msg)
        super("resque-exception - #{msg}")
      end

      # API request data structure.
      #
      # @return [Hash] data structure expected by the api.
      def api_request
        {
          'request' => {
            'parameters' => {
              'queue' => queue.to_s,
              'job_class' => payload['class'].to_s,
              'job_args' => payload['args'],
              'worker' => worker.to_s
            }
          },
          'application_environment' => {
            'env' => ENV.to_hash,
            'application_root_directory' => ENV['PWD']
          },
          'exception' => {
            'occurred_at' => Time.now.iso8601,
            'message' => "#{exception.class.name}: #{exception.message}",
            'backtrace' => Array(exception.backtrace),
            'exception_class' => exception.class.name
          },
          'rescue_block' => {
            'name' => 'Resque Failure'
          },
          'client' => {
            'name' => 'resque-exceptional',
            'version' => Resque::Failure::Exceptional::Version,
            'protocol_version' => 5
          }
        }
      end

    end
  end
end
